ACPs/77-reinventing-subnets/README.md

```diff
@@ -32,48 +32,194 @@ With no Elastic Subnets live on Mainnet, it is clear that Permissionless Subnets
+The following serialization is defined as a `ValidatorData`:
 
-To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).
+| Field | Type | Size |
+| -: | -: | -: |
+| `nodeID` | `[]byte` | 4 + len(`nodeID`) bytes |
+| `blsPublicKey` | `[48]byte` | 48 bytes |
+| `weight` | `uint64` | 8 bytes |
+| | | 60 + len(`nodeID`) bytes |
+
+
+The following serialization is defined as the `SubnetConversionData`:
+
+| Field | Type | Size |
+| -: | -: | -: |
+| `codecID`  | `uint16` | 2 bytes |
+| `subnetID` | `[32]byte` | 32 bytes |
+| `managerChainID` | `[32]byte` | 32 bytes |
+| `managerAddress` | `[]byte` | 4 + len(`managerAddress`) bytes |
+| `validators` | `[]ValidatorData` | 4 + sum(`validatorLengths`) bytes |
+| | | 74 + len(`managerAddress`) + len(`validatorLengths`) bytes |
+
+- `codecID` is the codec version used to serialize the payload, and is hardcoded to `0x0000`
+- `sum(validatorLengths)` is the sum of the lengths of `ValidatorData` serializations included in `validators`.
+- `subnetID` identifies the Subnet that is being converted (described further below).
+- `managerChainID` and `managerAddress` identify the validator manager for the given Subnet. This is the (blockchain ID, adress) tuple allowed to send Warp messages to modify the Subnet's validator set.
+- `validators` are the initial pay-as-you-go validators for the given Subnet.
+
+The `SubnetConversionMessage` is specified as an `AddressedCall` with `sourceChainID` set to the P-Chain ID, the `sourceAddress` set to an empty byte array, and a payload of:
+
+| Field | Type | Size |
+| -: | -: | -: |
+| `codecID` | `uint16` | 2 bytes |
+| `typeID` | `uint32` | 4 bytes |
+| `subnetConversionID` | `[32]byte` | 32 bytes |
+| | | 38 bytes |
+
+- `codecID` is the codec version used to serialize the payload, and is hardcoded to `0x0000`
+- `typeID` is the payload type identifier and is `0x00000000` for this message
+- `subnetConversionID` is the SHA256 hash of the `SubnetConversionData` from a given `ConvertSubnetTx`
 
-#### ConvertSubnetTx
+#### `RegisterSubnetValidatorMessage`
+
+The `RegisterSubnetValidatorMessage` is sent from validator managers to the P-Chain to inform the P-Chain of additions to Subnet's validator set.
+
+The following is the serialization of a `PChainOwner`:
+
+| Field | Type | Size |
+| -: | -: | -: |
+| `threshold` | `uint32` | 4 bytes |
+| `addresses` | `[][20]byte` | 4 + len(`addresses`) * 20 bytes |
```

Comment:
can you add explanation of these fields, similar to other messages. I'm wondering what threshold is, and why address is known to be 20 bytes, compared to others that are variable length.

Reply:
Added. 

All P-Chain addresses are represented as `ids.ShortID`, which are 20 bytes. I left that part out because I don't think it's super relevant here, similar to how we don't explain why BLS signatures are known to be 96 bytes, etc.